pyartcd/pyartcd/pipelines/build_microshift.py

```diff
@@ -0,0 +1,164 @@
+
+    def __init__(self, runtime: Runtime, group: str, assembly: str,
+                 ocp_build_data_url: str, logger: Optional[logging.Logger] = None):
+        self.runtime = runtime
+        self.group = group
+        self.assembly = assembly
+        self._logger = logger or runtime.logger
+        self._slack_client = self.runtime.new_slack_client()
+
+        # determines OCP version
+        match = re.fullmatch(r"openshift-(\d+).(\d+)", group)
+        if not match:
+            raise ValueError(f"Invalid group name: {group}")
+        self._ocp_version = (int(match[1]), int(match[2]))
+
+        # sets environment variables for Doozer
+        self._doozer_env_vars = os.environ.copy()
+        self._doozer_env_vars["DOOZER_WORKING_DIR"] = str(self.runtime.working_dir / "doozer-working")
+
+        if not ocp_build_data_url:
+            ocp_build_data_url = self.runtime.config.get("build_config", {}).get("ocp_build_data_url")
+        if ocp_build_data_url:
+            self._doozer_env_vars["DOOZER_DATA_PATH"] = ocp_build_data_url
+
+    async def run(self):
+        self._slack_client.bind_channel(self.group)
+        slack_response = await self._slack_client.say(f":construction: Build microshift for assembly {self.assembly} :construction:")
+        slack_thread = slack_response["message"]["ts"]
+
+        try:
+            await load_group_config(self.group, self.assembly, env=self._doozer_env_vars)
+            releases_config = await load_releases_config(Path(self._doozer_env_vars["DOOZER_WORKING_DIR"], "ocp-build-data"))
+            assembly_type = get_assembly_type(releases_config, self.assembly)
+            if assembly_type not in [AssemblyTypes.STANDARD, AssemblyTypes.CANDIDATE, AssemblyTypes.PREVIEW]:
+                raise ValueError(f"Building MicroShift for assembly type {assembly_type.value} is not currently supported.")
+            release_name = get_release_name(assembly_type, self.group, self.assembly, None)
+
+            # Rebases and builds microshift
+            version, release = self.generate_microshift_version_release(release_name)
+            nvrs = await self._rebase_and_build_rpm(version, release)
+
+            # https://mirror.openshift.com/pockets/microshift/4.10-el8/stream/x86_64/os/
+
+            # Prints example schema
+            click.secho("Build completes. Please update the assembly schema in releases.yaml to pin the following NVR(s) to the assembly:\n", fg="green")
+            for nvr in nvrs:
+                click.secho(f"\t{nvr}", fg="green")
+            example_schema = yaml.safe_dump(self._generate_example_schema(nvrs))
+            click.secho(f"\nExample schema:\n\n{example_schema}", fg="green")
+
+            # Sends a slack message
+            # TODO: We might want this job to auto create a PR to ocp-build-data.
+            await self._slack_client.say(f"Hi @release-artists , microshift for assembly {self.assembly} has been successfully built.", slack_thread)
```

Comment:
Do we want to notify even if `DRY_RUN` was specified?

Reply:
SlackClient class honors --dry-run option in its own implementation, so a slack message will not be actually sent in dry run mode.